### DISCUSSION TITLE
Snap!'s keyboard editor addon

### DISCUSSION THREAD

**Discussion by nexensys** [ScratchAddons/ScratchAddons#1629] [Ideas]
Hi there! I'm ErrorGamer2000, and I am working on a new addon. It is a keyboard editor similar to Snap!'s that will allow you to add blocks by typing them. 
### Current features: 
- Syntax highlighting
- Adds [this button](https://github.com/ErrorGamer2000/randomassets/blob/main/type.svg) to the editor above the zoom controls for opening/closing the editor.
- Block preview
- Some basic debugging*
<sub><sup>*Features in development</sup></sub>

### Future features:
- Multiple languages (ScratchScript[i.e. scratchblocks], JavaScript, Python)
- Autocomplete
- Add blocks by dragging them from preview, not just button.
- "Edit as text" in right-click menu allowing you to edit existing block stacks
- Match editor colors to current scratch theme

Anything else you want? Comment on this!

**Comment by retronbv** [comment]
> Multiple languages (ScratchScript[i.e. scratchblocks], JavaScript, Python)

So you’re gonna convert Python and JS to scratch blocks? That’s gonna be pretty hard. Do you have a working concept?

**Comment by retronbv** [comment]
How would that work tho? Do you have a code snippet?

**Comment by TheColaber** [comment]
I think in the meantime this should be turned into a discussion since you guys are discussing it a lot.

**Reply by Hans5958** [reply]
I think this should be an issue, instead as a discussion, since this is eventually going to be implemented to SA, and the author has an intent to do so. Doesn't matter if there is a lenghty discussion, in fact discussions are encouraged on issues before PRs are made. Something like small ideas is the one's that should be a discussion.

@ErrorGamer2000, if there is no objection, I advise you to make another issue linking this discussion post, since there is no way of us reverting it back to an issue.

**Reply by TheColaber** [reply]
I'm not saying I don't like this idea or anything, in fact, this is a great idea! But I also see that this is a very complex addon and for the time being there is going to be a lot of discussions. When the time comes that they finish this addon, they can 100% create an issue for it. I understand that discussions can be held in issues but it's better if discussions are held here.

**Reply by nexensys** [reply]
I agree, It helps keep it a little cleaner and more organized.

**Reply by Hans5958** [reply]
> 
> 
> I'm not saying I don't like this idea or anything, in fact, this is a great idea! But I also see that this is a very complex addon and for the time being there is going to be a lot of discussions. When the time comes that they finish this addon, they can 100% create an issue for it. I understand that discussions can be held in issues but it's better if discussions are held here.

I'm also not in the stance of rejecting the idea. I'm just saying that Discussions is quite a wrong place for this type of posts. Posts related to a feature/bug that is going to be implemented/fixed should go to the Issues tab, while general discussion about SA should be put on Discussions (see examples #1404, #817, #1628, #495). The Discussions feature is tailored for general discussions (features like upvotes, replies, etc), not something related to development like this (features like labeling, closing, etc)

Actually, when it is done, he will create an PR, not an issue.

**Comment by jeffalo** [comment]
this sounds really exciting, keep us posted

**Reply by nexensys** [reply]
Really? I will!

**Reply by TheColaber** [reply]
Yeah, this would be really cool! Let us know if you have a branch we can look at.

**Reply by nexensys** [reply]
So far, all I have is the injection for the elements I will need. I am still working on how to actually *add* the blocks to the workspace, and my editor code is in another html file because I am not *really* supposed to be on scratch right now. XD parents got mad because I spent most of the first semester of school working on scratch behind their backs... But I will put the current files I have in a branch!

**Reply by nexensys** [reply]
Not being able to visit the website does make it really hard to make an addon...

**Reply by nexensys** [reply]
Here you go: https://github.com/ErrorGamer2000/ScratchAddons

**Reply by nexensys** [reply]
**Question:** I have only actually been doing javascript/html/css for about four months now... I have gone through a lot of tutorials and seen a lot of code and can understand most of the things I see. Still, I am having a really hard time understanding how to add blocks to the workspace. I have read through every line of code in the devtools addon, and I picked up some things, but can't figure out the formatting I need to do... can someone help me understand this part? I am about to look through the blockly manual otherwise...

**Comment by nexensys** [comment]
Oooookay... I am forgoing the javascript and python at the moment, it is a massive job. I have actually gone and unminified the scratchblocks javascript and am using the key from that along with some of it's methods. I have also gotten the block preview to work.

**Comment by nexensys** [comment]
JUST HAD A MASSIVE BREAKTHOUGH! I had been attempting to make my own scratchblocks parser. I now know that I can just include scratchblocks and use scratchblocks.parse! That takes a massive amount of work away!

**Reply by retronbv** [reply]
ok!

**Comment by Endermaster3690** [comment]
i think this will do as a thing that might sorta give you a picture of what i imagine it as
made it in scratch's paint editor
![coding hkdfogphvdfpohdfoptkhd](https://user-images.githubusercontent.com/80355301/110525716-8c4cc780-80da-11eb-8d16-a81d5e475bfc.png)

**Reply by nexensys** [reply]
Here: 
![keyboard-editor](https://user-images.githubusercontent.com/70162741/110532800-e9e51200-80e2-11eb-9317-10d974083bee.png)

**Reply by Endermaster3690** [reply]
i was thinking of like, there were categories, then the text blocks that would insert where ever your cursor is, then the coding editor would be the programming, the block text would be colored as the category or a background. no don't make it the background cause selecting text would be weird. it would also show all scripts in the sprite. with yours also add script editing.

**Reply by nexensys** [reply]
I am working on a: Figuring out how to actually detect what blocks they are using, which is a massive job, I am having to use regexes, and b: trying to make it so that you can right-click on a block, select 'Edit Stack As Text' and it will automatically load the text for those blocks. I am also working on a syntax highlighter for it. But that is actually a good idea for a feature, I may add it eventually!

**Reply by Secret-chest** [reply]
> Here: 
> ![keyboard-editor](https://user-images.githubusercontent.com/70162741/110532800-e9e51200-80e2-11eb-9317-10d974083bee.png)
> 

What is that strange font?

**Reply by nexensys** [reply]
[Itim](https://fonts.google.com/specimen/Itim?preview.text_type=custom), from google fonts. I was using my other addon, the custom font addon (not officially added yet).

**Reply by Endermaster3690** [reply]
I think there should have to be an error custom block for incorrect coding. what i think of is that it prevents you from doing the typing until you make a custom block named ''error''. It gives you a pop up that tells you to make that block.

**Reply by nexensys** [reply]
... what??

**Reply by nexensys** [reply]
Can you explain that a little better?

**Reply by Mr-Programmer-777** [reply]
> [Itim](https://fonts.google.com/specimen/Itim?preview.text_type=custom), from google fonts. I was using my other addon, the custom font addon (not officially added yet).

Can I get that, please?

**Reply by nexensys** [reply]
The addon? Sure. It's in my fork.

**Reply by nexensys** [reply]
Here: https://github.com/ErrorGamer2000/ScratchAddons-1

**Reply by Endermaster3690** [reply]
> ... what?? Can you explain that a little better?

here ill paste what i said earlier than change it to be different, and i have a different idea. Here it is:

So it tells you errors in the code and it wont let you add that code until all errors fixed. It gives you a pop up that tells you to fix the errors if you add something bad and wont let you add the code till you do.

**Reply by nexensys** [reply]
Ooooh, yes. I did list a debugger as one of the features that I was going to add... It will be pretty basic to start, but will be upgraded. I am also hoping to have an autocomplete.

**Reply by Endermaster3690** [reply]
>  I am also hoping to have an autocomplete.

like it gives you some ideas on what to type while in the middle of typing something?

**Reply by nexensys** [reply]
Yes. Something like JSFiddle's autocomplete or google docs'.

**Reply by Endermaster3690** [reply]
> Yes. Something like JSFiddle's autocomplete or google docs'.

i was thinking of like, roblox studio recommendation thingamabobs?
did text above confuse you?

**Reply by nexensys** [reply]
Never played Roblox, what is thingamabobs?

**Reply by Endermaster3690** [reply]
Roblox is a kinda good kinda bad cause its ceo thing doesnt seem to be helping as far as i know.
A Thingamabob is just a thing i say its an unnecessary lengthening maybe of ''thing''.

**Reply by nexensys** [reply]
Okay, then.

**Reply by Endermaster3690** [reply]
well im guessing that this addon is getting accepted maybe?

**Reply by nexensys** [reply]
Most likely, in their own words.

**Reply by Endermaster3690** [reply]
if it is accepted i would be glad.

**Reply by nexensys** [reply]
It would decrease my production time by half...

**Reply by cobaltt7** [reply]
@Endermaster3690 they rarely decline addons

**Reply by nexensys** [reply]
Here is a little glimpse of my current code, for ONE BLOCK:
```js
{
    id: "motion_gotoxy",
    pattern: /go to x:(?: |)\((.*)\)(?: |)y:(?: |)\((.*)\)/,
    inputs: [
        {
            name: "X",
            type: "number",
            shape: "reporter",
        },
        {
            name: "Y",
            type: "number",
            shape: "reporter",
        },
    ],
};
```

**Reply by nexensys** [reply]
That regex tho... XD

**Reply by nexensys** [reply]
Oh, and the xml for that block:
```html
<block type="motion_gotoxy" id="motion_gotoxy">
    <value name="X">
        <shadow id="movex" type="math_number">
            <field name="NUM">0</field>
        </shadow>
    </value>
    <value name="Y">
        <shadow id="movey" type="math_number">
            <field name="NUM">0</field>
        </shadow>
    </value>
</block>
```

**Reply by Mr-Programmer-777** [reply]
Looks good. So, with that, can't you make 'glitched/hacked' blocks by editing the field types? E.g; set variable (varName) to 0, like the hacked create variable thing in scratch?

**Reply by nexensys** [reply]
I tried to change the menu types with Blockly, but the formatting is insane. The xml does not allow that, and the only thing I was able to do with an actual instance of Blockly was make a custom block the shape of an `if then` block that you could not put blocks in. (But it still works)

**Reply by nexensys** [reply]
I still have to devise a way to convert a custom block's proccode into a regex...

**Reply by Endermaster3690** [reply]
Doing custom blocks is going to be PAINFUL i think. How about we just finish scratch normal blocks, make a current beta build that doesnt do custom blocks.

**Reply by Mr-Programmer-777** [reply]
> I tried to change the menu types with Blockly, but the formatting is insane. The xml does not allow that, and the only thing I was able to do with an actual instance of Blockly was make a custom block the shape of an `if then` block that you could not put blocks in. (But it still works)

What about HTML? I heard XML's strict.

**Reply by nexensys** [reply]
No, Blockly (at least in scratch) only takes xml. If you can find a better way to add the blocks, it would be appreciated. I tried looking through the devtools addon, but can't really find it. I have been reading over it for a month and still can't find it...

**Reply by qbjl** [reply]
> Here:
> ![keyboard-editor](https://user-images.githubusercontent.com/70162741/110532800-e9e51200-80e2-11eb-9317-10d974083bee.png)

Cool, but I was thinking like the custom "Variables" tab you can toggle, where it has a whole editor instead of in front of the scratchblocks

**Comment by Mr-Programmer-777** [comment]
Honestly, this sounds amazing. Do you plan on making it a tad like [tosh](https://tosh.blob.codes/)? Otherwise, this could lead to code from JS or Python being executed, like arrays or other things. And, implementing the whole of JS and Python could even let users get scratch login tokens, which I think is VERY bad. Still, I have complete support if it was to work like tosh. Otherwise, I would love this to happen, but the 5,000,000 vunerabilities and cons outweigh it.

**Reply by nexensys** [reply]
YES, omg I had forgotten about that... that might help me some... But also no, I will not be running the blocks only adding them to the workspace from the text. I am going to add them via `Blockly.Xml` as such:
```js
Blockly.Xml.domToWorkspace(Blockly.Xml.textToDom('<xml><!--Block code here--></xml>'), Blockly.getMainWorkspace())
```

**Comment by Endermaster3690** [comment]
We need your custom fonts addon in SA. however it should be able to change fonts of a lot of categorized text.

**Reply by nexensys** [reply]
They shut it down...

**Comment by Endermaster3690** [comment]
hows the work going?

**Reply by nexensys** [reply]
...slowly. I don't have much time to work on it right now. School is starting to catch up to me. I most likely will not be able to work on it any more until week after next, and I will be lucky to get 2 hours in at that time...

**Reply by Endermaster3690** [reply]
aw man. Well do as much as you can!

oh no that rhymed

**Reply by nexensys** [reply]
ah, well. Nobody's perfect...😂

**Reply by nexensys** [reply]
Here, I will make a new repo so you can see my current progress...

**Reply by nexensys** [reply]
[Here](https://github.com/ErrorGamer2000/Scratch-Keyboard-Editor) you go!

**Comment by nexensys** [comment]
XD I just found that I have to do almost no translations for this, since I can just do
```js

export default async function ({addon, global, console, msg}) {
    const fullBlockly = await addon.tab.traps.getBlockly();
    const msgs = fullBlockly.ScratchMsgs.locales[msg.locale];
}
```
